2019-11-storybook-for-web-components-on-steroids/README.md

```diff
@@ -0,0 +1,497 @@
+Knowing that it makes sense that there are 2 separate builds.
+
+But why is there a build step at all? Why would storybook have such setup?
+
+Let's see what is needed to allow for some code to be run and worked on in multiple browsers.
+
+### Excursion shipping code based on browser capabilities
+
+Let's have a small example where we are using [private class fields](https://github.com/tc39/proposal-class-fields).
+This feature is currently at stage 3 and only available in Chrome.
+
+```js
+// index.js
+import { MyClass } from './MyClass.js';
+
+const inst = new MyClass();
+inst.publicMethod();
+
+// MyClass.js
+export class MyClass {
+  #privateField = 'My Class with a private field';
+
+  publicMethod() {
+    document.body.innerHTML = this.#privateField;
+    debugger;
+  }
+}
+```
+
+We deliberately put a debugger breakpoint in there to see the actual code the browser is executing.
+
+Let's see how webpack with a few babel plugins handles it. ([see full config](./EsDevServer-vs-WebpackDevServer/webpack.config.js))
+
+```js
+__webpack_require__.r(__webpack_exports__);
+/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "MyClass", function() { return MyClass; });
+function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
+function _defineProperties(target, props) {
+// ... more helper functions
+
+var MyClass =
+/*#__PURE__*/
+function () {
+  function MyClass() {
+    _classCallCheck(this, MyClass);
+
+    _privateField.set(this, {
+      writable: true,
+      value: 'My Class with a private field'
+    });
+  }
+
+  _createClass(MyClass, [{
+    key: "publicMethod",
+    value: function publicMethod() {
+      document.body.innerHTML = _classPrivateFieldGet(this, _privateField);
+      debugger;
+    }
+  }]);
+
+  return MyClass;
+}();
+
+var _privateField = new WeakMap();
+```
+
+Wow that is quite some code 🙈 and it does not really look like the code written 😱
+
+what happens? in a typical webpack & babel setup your code gets compiled down to es5 in order to be able to run the code also on older browser like IE11.
+
+However, you may ask how often do I actually run my app in an older browser?
+
+A typical developer should probably develop ~90% on a modern browser and ~10% on older browsers to make sure everything still works in order.
+At least we hope you have such a nice workflow 🤗
+
+So the question is why compile, ship, debug and work with this "strange" code 100% of the time if it's only needed for 10%?
+Could we do better?
+
+Let's see how `es-dev-server` handles it by opening the same file on chrome.
+
+```js
+export class MyClass {
+  #privateField = 'My Class with a private field';
+
+  publicMethod() {
+    document.body.innerHTML = this.#privateField;
+    debugger;
+  }
+}
+```
+
+It looks exactly as the original code - because it is. The code as is was fully capable to run in chrome without any adjustments.
```

Comment:
```suggestion
It looks exactly as the original code - because it is. The code as is was fully capable of running in chrome without any adjustments.
```